src/api/bulk2.ts

```diff
@@ -0,0 +1,1000 @@
+    ].join('/');
+
+    return new BulkApiV2(this.connection, { responseType }).request<T>({
+      ...request,
+      url: path ? baseUrl + path : baseUrl,
+    });
+  }
+}
+
+class JobDataV2<
+  S extends Schema,
+  Opr extends IngestOperation
+> extends Writable {
+  readonly job: IngestJobV2<S, Opr>;
+  readonly uploadStream: Serializable;
+  readonly downloadStream: Parsable;
+  readonly dataStream: Duplex;
+  result: any;
+
+  /**
+   *
+   */
+  constructor(options: CreateJobDataV2Options<S, Opr>) {
+    super({ objectMode: true });
+
+    const createRequest = options.createRequest;
+
+    this.job = options.job;
+    this.uploadStream = new Serializable();
+    this.downloadStream = new Parsable();
+
+    const converterOptions = { nullValue: '#N/A' };
+    const uploadDataStream = this.uploadStream.stream('csv', converterOptions);
+    const downloadDataStream = this.downloadStream.stream(
+      'csv',
+      converterOptions,
+    );
+
+    this.dataStream = concatStreamsAsDuplex(
+      uploadDataStream,
+      downloadDataStream,
+    );
+
+    this.on('finish', () => this.uploadStream.end());
+
+    uploadDataStream.once('readable', () => {
+      try {
+        // pipe upload data to batch API request stream
+        const req = createRequest({
+          method: 'PUT',
+          path: `/${this.job.jobInfo?.id}/batches`,
+          headers: {
+            'Content-Type': 'text/csv',
+          },
+          responseType: 'application/json',
+        });
+
+        (async () => {
+          try {
+            const res = await req;
+            this.emit('response', res);
+          } catch (err) {
+            this.emit('error', err);
+          }
+        })();
+
+        uploadDataStream.pipe(req.stream());
+      } catch (err) {
+        this.emit('error', err);
+      }
+    });
+  }
+
+  _write(record_: Record, enc: BufferEncoding, cb: () => void) {
+    const { Id, type, attributes, ...rrec } = record_;
+    let record;
+    switch (this.job.jobInfo.operation) {
+      case 'insert':
+        record = rrec;
+        break;
+      case 'delete':
+      case 'hardDelete':
+        record = { Id };
+        break;
+      default:
+        record = { Id, ...rrec };
+    }
+    this.uploadStream.write(record, enc, cb);
+  }
+
+  /**
+   * Returns duplex stream which accepts CSV data input and batch result output
+   */
+  stream() {
+    return this.dataStream;
+  }
+
+  /**
+   * Execute batch operation
+   */
+  execute(input?: string | Record[] | Readable) {
+    if (this.result) {
+      throw new Error('Data can only be uploaded to a job once.');
+    }
+
+    this.result = new Promise<void>((resolve, reject) => {
+      this.once('response', () => resolve());
+      this.once('error', reject);
+    });
+
+    if (is.nodeStream(input)) {
+      // if input has stream.Readable interface
+      input.pipe(this.dataStream);
+    } else {
+      if (Array.isArray(input)) {
+        for (const record of input) {
```

Comment:
I don't like that this mutates the input.  Imagine being a jsforce consumer, and you passed in your options, passed down through a bunch of stuff, and we changed all the booleans to string on your data.

Reply:
maybe construct a new object that gets written from `Object.entries(Object.fromEntries(record).map("<boolean=>string stuff>"))

Reply:
fixed by using `structuredClone` in both bulk modules.